rustls/src/crypto/ring.rs

```diff
@@ -95,3 +108,123 @@ impl ProducesTickets for AeadTicketer {
         Some(out)
     }
 }
+
+/// An in-progress key exchange. This has the algorithm,
+/// our private key, and our public key.
+pub struct RingKeyExchange {
+    skxg: &'static SupportedKxGroup,
+    privkey: EphemeralPrivateKey,
+    pub(crate) pubkey: ring::agreement::PublicKey,
+}
+
+impl Debug for RingKeyExchange {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.skxg.fmt(f)
+    }
+}
+
+impl KeyExchange for RingKeyExchange {
+    type SupportedGroup = SupportedKxGroup;
+
+    fn all_supported_groups() -> &'static [&'static Self::SupportedGroup] {
```

Comment:
I like to lay out methods in a sort of top-down order. Constructors come first, trivial functions come last. More complex before least complex. Dependent before dependee. `&mut self` receivers before `&self` receivers.